Entity Analytics Portfolio updates (#5994)
elastic/docs-content: solutions/security/advanced-entity-analytics/entity-store.md

Reviewer: We should add a disclaimer for low-resource deployments.
https://elastic.slack.com/archives/C09U0KU22RY/p1777186936773939?thread_ts=1777035895.684139&cid=C09U0KU22RY

{applies_to}`stack: preview 9.4+` {applies_to}`serverless: planned` Entity relationships sourced from the entity store — such as access patterns, dependencies, and resolution links — are visible in the entity details flyout's [Graph View](/solutions/security/advanced-entity-analytics/view-entity-details.md#visualizations) tab. Entities that appear in both the entity store and in raw events are rendered as a single deduplicated node in the graph.

When the entity store is enabled, the following resources are created for the active space:

:::::{applies-switch}

::::{applies-item} { stack: ga 9.4+, serverless: planned }
* A latest entity alias, `entities-latest-<space-id>`, backed by the concrete index `.entities.v2.latest.security_<space-id>-<mapping_version>`. Query this alias to retrieve the current state of all entities in the entity store.
* History snapshot indices, `.entities.v2.history.security_<space-id>.<timestamp>`, which store daily snapshots of entity data and enable [historical analysis](/solutions/security/advanced-entity-analytics/view-analyze-risk-score-data.md#historical-entity-analysis) of entity attributes over time.

:::{note}

> Starting in 9.4, the entity store uses {{esql}}-based LOOKUP JOIN queries instead of {{es}} transforms and moves from transform-based indices (`.entities.v1.*`) to ES|QL-based indices (`.entities.v2.*`). When you upgrade from a previous version, existing transforms, enrich policies, and ingest pipelines are removed. Your existing index data is retained.

Reviewer 2: ```suggestion
Starting in 9.4, the entity store uses {{esql}}-based LOOKUP JOIN queries instead of {{es}} transforms and moves from transform-based indices (`.entities.v1.*`) to ES|QL-based indices (`.entities.v2.*`). When you upgrade from a previous version, existing transforms, enrich policies, and ingest pipelines are removed. Your existing index data is retained. After the entity store is enabled, historical Entity data from logs within the last 3 hours will be extracted.
```

> # Entity store [entity-store]

Reviewer: From the docs:
> Clearing entity store data does not delete your source data, assigned entity risk scores, or asset criticality assignments.

I don’t think this is true anymore. Clearing Entity Store data will require new asset criticality assignments. For risk score, the Risk Engine will recalculate it when it runs (~1h), and the updated/new score will be written back into the Entity Store in case the entity was already created.

@jaredburgettelastic keep me honest here.

Reviewer 2: @uri-weisman that's correct, asset criticality no longer has its own index, and is fully reliant upon what is in the entity store.

Reviewer: @natasha-moore-elastic can we modify it to something like:
_Clearing Entity Store data does not delete any source data. However, asset criticality assignments will need to be reapplied, and risk scoring will run again for the new entities repopulated into the store.._

{applies_to}`stack: preview 9.4+` {applies_to}`serverless: preview` Entity relationships sourced from the entity store — such as access patterns, dependencies, and resolution links — are visible in [Graph Visualization](/solutions/security/advanced-entity-analytics/view-entity-details.md#visualizations) from the entity details flyout. Entities that appear in both the entity store and in raw events are rendered as a single deduplicated node in the graph.

When the entity store is enabled, the following resources are created for the active space:

:::::{applies-switch}

::::{applies-item} { stack: ga 9.4+, serverless: ga }
* A latest entity alias, `entities-latest-<space-id>`, backed by the concrete index `.entities.v2.latest.security_<space-id>-<mapping_version>`. Query this alias to retrieve the current state of all entities in the entity store.
* History snapshot indices, `.entities.v2.history.security_<space-id>.<timestamp>`, which store daily snapshots of entity data and enable [historical analysis](/solutions/security/advanced-entity-analytics/view-analyze-risk-score-data.md#historical-entity-analysis) of entity attributes over time.

:::{note}

> Entity store v2 uses {{esql}}-based LOOKUP JOIN queries instead of {{es}} transforms. When v2 is installed, v1 transforms, enrich policies, and ingest pipelines are removed. Your v1 index data is retained.

Reviewer 3: Here, the `applies_to` tag is doing the versioning work, instead of the prose.

```suggestion
:applies_to: {"stack": "ga 9.4+", "serverless": "ga"}
Entity store uses {{esql}}-based LOOKUP JOIN queries instead of {{es}}
transforms. When upgraded, existing transforms, enrich policies, and ingest
pipelines are removed. Your existing index data is retained.
```

Author: Thanks @nastasha-solomon! The content is already under `{applies-item} { stack: ga 9.4+, serverless: ga }` but I will rephrase based on your suggestions, to remove v1/v2 references

:::

:::{warning}

> Entity store v2 replaces per-type v1 indices with a single shared latest alias. Update any direct queries or automations that reference `.entities.v1.latest.security_user_*`, `.entities.v1.latest.security_host_*`, or `.entities.v1.latest.security_service_*` to use `entities-latest-<space-id>` instead. The v1 API routes are deprecated in 9.4.

Reviewer 3: Same as above - recommend using the `applies_to` tags to convey versioning. 

```suggestion
:applies_to: stack: deprecated 9.4
Entity store moved from transform-based indices (`.entities.v1.*`) to
{{esql}}-based indices (`.entities.v2.*`), which are exposed through the shared
`entities-latest-<space-id>` alias. Update any direct queries or automations
that reference `.entities.v1.latest.security_user_*`,
`.entities.v1.latest.security_host_*`, or `.entities.v1.latest.security_service_*`
to use `entities-latest-<space-id>` instead. The previous API routes are deprecated.
```

* {applies_to}`stack: ga 9.2+` Snapshot indices (`.entities.v1.history.<ISO_date>.*`) that store daily snapshots of entity data, enabling [historical analysis](/solutions/security/advanced-entity-analytics/view-analyze-risk-score-data.md#historical-entity-analysis) of attributes over time.
* {applies_to}`stack: ga 9.2+` Reset indices (`.entities.v1.reset.*`) that ensure entity timestamps are updated correctly in the latest index, supporting accurate time-based queries and future data resets.
::::

:::::

## Enable entity store [enable-entity-store]

::::{applies-switch}

:::{applies-item} { stack: ga 9.4+, serverless: ga }
The entity store is automatically enabled when you turn on the risk scoring engine. In the default {{kib}} space, both are enabled automatically. In non-default spaces, you must enable them manually:

1. Find the **Entity Analytics** management page in the navigation menu or by using the [global search field](/explore-analyze/find-and-organize/find-apps-and-objects.md).
2. Turn the toggle on.

:::{note}

> If you've upgraded from a previous version, and entity store v1 was installed in any space, it's automatically migrated to v2 after the upgrade. Your v1 index data is retained.

Reviewer 3: ```suggestion
:applies_to: stack: ga 9.4
If you've upgraded from a previous version, and the entity store was installed in any space, it's automatically migrated after the upgrade. Your index data is retained.
```

Author: This content is already under `{applies-item} { stack: ga 9.4+, serverless: ga }`, so I don't think additional tags are needed here, but I'll rephrase to remove v1/v2 references

* Asset criticality data
* Entity risk score data

The entity store can hold any entity type observed by {{elastic-sec}}. It allows you to view and query select entities represented in your indices without needing to perform real-time searches of observable data. The entity store extracts entities from all indices in the {{elastic-sec}} [default data view](../get-started/data-views-elastic-security.md#default-data-view-security).

{applies_to}`stack: ga 9.4+` {applies_to}`serverless: ga` [Entity resolution](/solutions/security/advanced-entity-analytics/entity-resolution.md) is built on top of the entity store. It links multiple entity records representing the same real-world identity into a resolution group, consolidating their risk scores into a single view.

> {applies_to}`stack: preview 9.4+` {applies_to}`serverless: preview` Entity relationships sourced from the entity store — such as access patterns, dependencies, and resolution links — are visible in the entity details flyout's [Graph View](/solutions/security/advanced-entity-analytics/view-entity-details.md#visualizations) tab. Entities that appear in both the entity store and in raw events are rendered as a single deduplicated node in the graph.

Reviewer 2: These currently show up as `serverless: preview`, which isn't correct as of yet, as the 9.4 release will not happen at the same time as the Serverless release. I believe that we should instead render these as `serverless: planned`, though I'm unsure if that string the proper way to do this in the docs. 

This would apply to most instances in this PR where we do `serverless: preview`.

Reviewer 4: ```suggestion
{applies_to}`stack: preview 9.4+` {applies_to}`serverless: planned` Entity relationships sourced from the entity store — such as access patterns, dependencies, and resolution links — are visible in the entity details flyout's [Graph View](/solutions/security/advanced-entity-analytics/view-entity-details.md#visualizations) tab. Entities that appear in both the entity store and in raw events are rendered as a single deduplicated node in the graph.
```

When the entity store is enabled, the following resources are created for the active space:

:::::{applies-switch}

> ::::{applies-item} { stack: ga 9.4+, serverless: ga }

Reviewer 4: ```suggestion
::::{applies-item} { stack: ga 9.4+, serverless: planned }
```

The entity store can hold any entity type observed by {{elastic-sec}}. It allows you to view and query select entities represented in your indices  without needing to perform real-time searches of observable data. The entity store extracts entities from all indices in the {{elastic-sec}} [default data view](../get-started/data-views-elastic-security.md#default-data-view-security).

When the entity store is enabled, the following resources are created for the active space:

:::::{applies-switch}

::::{applies-item} { stack: ga 9.4+, serverless: ga }
* A latest entity alias, `entities-latest-<space-id>`, backed by the concrete index `.entities.v2.latest.security_<space-id>-<mapping_version>`. Query this alias to retrieve the current state of all entities in the entity store.

> * An updates data stream, `.entities.v2.updates.security_<space-id>` (alias: `entities-updates-<space-id>`), which records incremental entity updates as they are extracted from source data.

Reviewer: The update index is for internal use. I mentioned it because it is created for every space, and we do not want customers to modify or delete it.
Also, the description "which records incremental entity updates as they are extracted from source data" is not entirely accurate. I would instead describe it as an internal index used by the Entity Store architecture.

Author: Thanks @uri-weisman, if this is for internal use and we don't want users interacting with it, then I'd vote for removing it from the docs.

* History snapshot indices, `.entities.v2.history.security_<space-id>.<timestamp>`, which store daily snapshots of entity data and enable [historical analysis](/solutions/security/advanced-entity-analytics/view-analyze-risk-score-data.md#historical-entity-analysis) of entity attributes over time.

:::{note}

> Entity Store v2 uses {{esql}}-based LOOKUP JOIN queries instead of {{es}} ENRICH policies. When v2 is installed, v1 transforms, enrich policies, and ingest pipelines are removed. Your v1 index data is retained.

Reviewer: ```suggestion
Entity Store v2 uses {{esql}}-based LOOKUP JOIN queries instead of {{es}} transforms. When v2 is installed, v1 transforms, enrich policies, and ingest pipelines are removed. Your v1 index data is retained.
```

:::

::::

> ## Supported integrations [entity-store-integrations]

Reviewer 5: @uri-weisman don't we want to mention that any integration following ecs standards for `user`, `host`, and `event` fields should be supported?

Reviewer: I wonder it's not a bit of confusing, because host entities needs only `host.*` fields, while user entities needs at least user + host/user+event.

Maybe something like:

`The Entity Store creates user and host entities from data in supported source indices (mainly the Security Default Data View) when the incoming events include the ECS fields needed to identify those entities. In general, integrations that populate standard ECS identity fields such as `user.*`, `host.*`, and related `event.*` context can contribute to entity creation, as long as the data contains enough information for the Entity Store to reliably recognize and build the entity.
`

@natasha-moore-elastic wdyt?

Author: Updated in [d8ab051](https://github.com/elastic/docs-content/pull/5994/commits/d8ab0518e29fa9ed67854b1b1ec81ca03eb4c904).

* [Active Directory Entity Analytics](integration-docs://reference/entityanalytics_ad.md)
* [Microsoft Entra ID Entity Analytics](integration-docs://reference/entityanalytics_entra_id.md)
* [Okta Entity Analytics](integration-docs://reference/entityanalytics_okta.md)
* [Google Workspace](integration-docs://reference/google_workspace.md)
* [Microsoft 365](integration-docs://reference/o365.md)
* [AWS CloudTrail](integration-docs://reference/aws/cloudtrail.md)

**Endpoint and host sources:**

* [{{elastic-defend}}](integration-docs://reference/endpoint/index.md)
* [CrowdStrike](integration-docs://reference/crowdstrike.md)
* [SentinelOne](integration-docs://reference/sentinel_one.md)
* [Microsoft Defender for Endpoint](integration-docs://reference/microsoft_defender_endpoint.md)

## Troubleshoot entity store performance [entity-store-troubleshoot]
```yaml {applies_to}
stack: ga 9.4+
serverless: ga
```

The entity store runs scheduled log extraction to keep entity data up to date.

To determine whether log extraction is slow or unhealthy, check the **Engine Status** tab or query the Entity store status API.

A process might be **slow** if:

* New entities are not appearing as expected.

> * The last successful execution does not appear to advance (`lastExecutionTimestamp`).

Reviewer 5: This information is avaialble via the API only, not via the UI

Author: Updated in [d8ab051](https://github.com/elastic/docs-content/pull/5994/commits/d8ab0518e29fa9ed67854b1b1ec81ca03eb4c904).

A process might be **unhealthy** if:

* The engine enters an `error` state.
* Component health indicators are degraded.
* Extraction appears stalled and no forward progress is visible.

If log extraction appears slow, you can modify the following log extraction configuration settings to balance freshness, coverage, and query cost.

#### `frequency`

Use `frequency` to control how often extraction runs.

* Decrease frequency if extraction is healthy but too resource-intensive and {{es}} CPU utilization is too high. The minimum supported value is `30s`.

#### `docsLimit`

Use `docsLimit` to control how many entities can be processed in one extraction page.

> * Lower it if queries are too heavy or time-consuming.

Reviewer 5: Lower if kibana is consuming too much kibana memory. This doesn't have impact on ES query speed or mem/cpu

Author: Updated in [d8ab051](https://github.com/elastic/docs-content/pull/5994/commits/d8ab0518e29fa9ed67854b1b1ec81ca03eb4c904).
